### PR TITLE
Jwt auth

### PR DESCRIPTION
**Pull request by jdelamare**
This branch brings in a different authentication mechanism. This closes the more documented issue #160 as well as the generic one #168 (it can be further specified later). 

The auth works like this: 
* frontend sends JWT in querystring
* backend checks it in custom middleware
* websocket is either established or it is immediately closed with error 1002 indicating a protocol error.

JWT authentication uses the `DJANGO_SECRET_KEY` to check the signature, verifies that the user is in the database, and checks the timestamp.

`AuthMiddlewareStack` ASGI wrapper is no longer needed and was replaced by custom middleware helpfully outlined in the [channels docs](https://channels.readthedocs.io/en/latest/topics/authentication.html#custom-authentication).

### CODE REVIEW

File: web/src/index.ts [hunk 1]
```diff
@@ -115,7 +115,7 @@ const sendSocketCommand = wat => {
   if (wat.cmd == 'connect') {
     // console.log("connecting!");
     // let socket = new WebSocket(wat.address, wat.protocol);
-    let socket = new WebSocket(wat.address);
+    let socket = new WebSocket(wat.address, "asdf");
```
Comment (bgins) [inline web/src/index.ts:118]: The frontend is producing the following error when connecting over the websocket:

```
Error during WebSocket handshake: Sent non-empty 'Sec-WebSocket-Protocol' header but no response was received
```

It looks like the backend isn't expecting the `Sec-WebSocket-Protocol` header. The frontend connects appropriately when the `"asdf"` param is removed.

**Review by bgins — CHANGES_REQUESTED** [review]
A couple of changes to request here and a question. 

- The frontend is sending a `Sec-WebSocket-Protocol` header and it looks like we may be able to remove it
- There are some merge conflicts when pulling in the latest from `upgrade/staging` to resolve

It looks like running the Redis server is mandatory with these changes? The backend is failing to set up the websocket connection when it is not running.

File: web/src/index.ts (hunk 1 above)
Comment (jdelamare) [reply web/src/index.ts:118]: Yeah that's a big mistake on my part. That should have been removed before making the PR but it was overlooked. Remnants of when we were playing around trying to find the correct place to put the JWT.

I imagine that error will go away after this correction is made.

**Comment by jdelamare** [comment]
I'm not sure that the redis server needs to be running, as I haven't seen that error on my side yet. I'll play around further to check if that's the case.

File: ereadingtool/settings.py [hunk 2]
```diff
@@ -102,7 +108,6 @@
 DEV = False
 
 ALLOWED_HOSTS = ['0.0.0.0',
-                 'localhost',
```
Comment (bgins) [inline ereadingtool/settings.py:105]: Could we add `localhost` back in here?

File: web/src/Api/Config.elm [hunk 3]
```diff
@@ -31,9 +31,10 @@ init maybeConfig =
         Just config ->
             config
 
+        -- Here the URL has been changed for local development
         Nothing ->
             Config
-                { restApiUrl = "https://api.stepstoadvancedreading.org"
+                { restApiUrl = "localhost:8000"
```
Comment (bgins) [inline web/src/Api/Config.elm:37]: We may want to change this back to make the default case the production API endpoint.

**Review by bgins — APPROVED** [review]
Looks good! A couple of tiny changes to make, but it's all working. Go ahead and merge it in once you make those changes.

**Pull request merged**